### ISSUE TITLE
MNIST sample does not achieve expected accuracy (enable training CPU only)

### ISSUE THREAD

**Issue by georgen117** [microsoft/onnxruntime#6073]
**Describe the bug**
I followed instructions posted by @liqunfu in #3706 to successfully run the provided mnist training sample.

> I attached the mnist test data and models. Please run with provided command line. As you can see, to work with cpp API, you > need to follow the mnist example to build the graph with loss output so that backprop graph can be constructed.
>
> mnist.zip
> ./onnxruntime_training_mnist --model_name ~/mnist/mnist_gemm --train_data_dir ~/mnist/mnist_data/

Running `./onnxruntime_training_mnist --model_name ~/mnist/mnist_gemm --train_data_dir ~/mnist/mnist_data/` will result in a model trained to **89% accuracy**.

However, it will also print the following error for each input:

```
[W:onnxruntime:Default, graph.cc:1066 onnxruntime::Graph::Graph] Initializer <INPUT_NAME> appears in graph inputs and will not be treated as constant value/weight. This may prevent some of the graph optimizations, like const folding. Move it out of graph inputs if there is no need to override it, by either re-generating the model with latest exporter/converter or with the tool onnxruntime/tools/python/remove_initializer_from_input.py.
```

Following the recommendation of the warning, the `remove_initializer_from_input.py` is was run on the mnist_gemm model.

```
cd ~/mnist
python ~/onnxruntime/tools/python/remove_initializer_from_input.py --input mnist_gemm.onnx --output mnist_gemm2.onnx
```

Re-running with the updated model
```
 ./onnxruntime_training_mnist --model_name ~/mnist/mnist_gemm2 --train_data_dir ~/mnist/mnist_data/
```

This will result in a model trained to **24% accuracy**.

Until just a while ago both the `mnist_gemm.onnx` and `mnist_gemm2.onnx` models gave the same behavior when run with the `onnruntime_training_mnist` sample.

Using git bisect I was able to find the first commit that the behavior changed.

```
commit 24016a517bfa4d9dbc85768e045c95718221dc94
Author: Zhang Lei <zhang.huanning@hotmail.com>
Date:   Thu Nov 5 22:35:24 2020 -0800

    Prepacking in Gemm with merged logic for Matmul and Gemm on PackingB. (#5693)

    Prepacking in Gemm with merged logic for Matmul and Gemm on PackingB.
```

**Urgency**
none.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Windows 10** and **Linux Ubuntu 10.05.5 LTX**
- ONNX Runtime installed from (source or binary): **source**
- ONNX Runtime version: 24016a517bfa4d9dbc85768e045c95718221dc94
- Python version: **3.8.3**
- Visual Studio version (if applicable): **Microsoft Visual Studio Professional 2019 Version 16.7.7**
- GCC/Compiler version (if compiling from source): **GCC version 7.5.0**
- CUDA/cuDNN version: **N/A**
- GPU model and memory: **N/A**

**To Reproduce**
- Describe steps/code to reproduce the behavior.
  - Download mnist.zip from here:
  - https://github.com/microsoft/onnxruntime/files/4553416/mnist.zip
  - unzip to a known location
  - cd into the folder
  - run remove_initializer_from_input.py on mnist_gemm.onnx model
     i.e. `python ~/onnxruntime/tools/python/remove_initializer_from_input.py --input mnist_gemm.onnx --output mnist_gemm2.onnx`
  - from local onnxruntime repository
  - checkout master branch or 24016a517bfa4d9dbc85768e045c95718221dc94
  - build onnxruntime with training enabled. 
   `.\build.bat --config RelWithDebInfo --parallel  --build_shared_lib --cmake_generator "Visual Studio 16 2019" --enable_training` for windows
   `./build.sh --config RelWithDebInfo --parallel  --build_shared_lib --enable_training` for linux
  - cd into the build folder
  - run ` ./onnxruntime_training_mnist --model_name ~/mnist/mnist_gemm2 --train_data_dir ~/mnist/mnist_data/`
   

- Attach the ONNX model to the issue (where applicable) to expedite investigation.

https://github.com/microsoft/onnxruntime/files/4553416/mnist.zip

**Expected behavior**
The model to output accuracy of 89%
```
Evaluating the final model on the test set.
Test data range: [0 - 99)
Step: 0, #examples: 100, #correct: 89, precision: 0.8900, loss: 0.0035
```

**Additional context**
I am seeing the same behavior when running `orttraining/pytorch_frontend_examples/mnist_training.py` sample.  However, to get this sample running using CPU training requires using the --build_wheel option and modifying the mnist_training.py sample to work without CUDA. As well as some other steps.  If desired can provide those instructions.

**Comment by pranav-prakash** [comment]
I ran into this issue as well. After commit 24016a5 `Gemm` uses pre-packed weights for initializers that are constant; how does this interact with the weight update from `SGDOptimizer`? Once `SGDOptimizer` assigns the updated weights, the pre-packed weights should no longer be used. Have not debugged too deeply, but is there any mechanism currently in place to enforce this?

Alternatively instead of discarding pre-packed weights upon a weight update, perhaps pre-packing should not be done in the first place for weights that are to be trained.

**Comment by baijumeswani** [comment]
closing this since this seems old and `ORTTrainer` has been deprecated.
